File: Graph/GraphAM.js
var GraphAM = /** @class */ (function () {
    function GraphAM(cnt) {
        if (this.count === undefined)
            this.count = 0;
        if (this.adj === undefined)
            this.adj = null;
        this.count = cnt;
        this.adj = (function (dims) { var allocate = function (dims) { if (dims.length == 0) {
            return 0;
        }
        else {
            var array = [];
            for (var i = 0; i < dims[0]; i++) {
                array.push(allocate(dims.slice(1)));
            }
            return array;
        } }; return allocate(dims); })([this.count, this.count]);
    }
    GraphAM.prototype.addDirectedEdge = function (src, dst, cost) {
        this.adj[src][dst] = cost;
    };
    GraphAM.prototype.addUndirectedEdge = function (src, dst, cost) {
        this.addDirectedEdge(src, dst, cost);
        this.addDirectedEdge(dst, src, cost);
    };
    GraphAM.prototype.print = function () {
        for (var i = 0; i < this.count; i++) {
            {
                console.info("Node index [ " + i + " ] is connected with : ");
                for (var j = 0; j < this.count; j++) {
                    {
                        if (this.adj[i][j] !== 0)
                            console.info(j + " ");
                    }
                    ;
                }
                console.info("");
            }
            ;
        }
    };
    GraphAM.main1 = function () {
        var graph = new GraphAM(4);
        graph.addUndirectedEdge(0, 1, 1);
        graph.addUndirectedEdge(0, 2, 1);
        graph.addUndirectedEdge(1, 2, 1);
        graph.addUndirectedEdge(2, 3, 1);
        graph.print();
    };
    GraphAM.dijkstra = function (gph, source) {
        var previous = (function (s) { var a = []; while (s-- > 0)
            a.push(0); return a; })(gph.count);
        var dist = (function (s) { var a = []; while (s-- > 0)
            a.push(0); return a; })(gph.count);
        var visited = (function (s) { var a = []; while (s-- > 0)
            a.push(false); return a; })(gph.count);
        for (var i = 0; i < gph.count; i++) {
            {
                previous[i] = -1;
                dist[i] = 2147483647;
                visited[i] = false;
            }
            ;
        }
        dist[source] = 0;
        previous[source] = -1;
        var comp = new GraphAM.EdgeComparator();
        var queue = (new PriorityQueue(100, (comp)));
        var node = new GraphAM.Edge(source, 0);
        queue.add(node);
        while ((queue.isEmpty() !== true)) {
            {
                node = queue.peek();
                queue.remove();
                source = node.dest;
                visited[source] = true;
                for (var dest = 0; dest < gph.count; dest++) {
                    {
                        var cost = gph.adj[source][dest];
                        if (cost !== 0) {
                            var alt = cost + dist[source];
                            if (dist[dest] > alt && visited[dest] === false) {
                                dist[dest] = alt;
                                previous[dest] = source;
                                node = new GraphAM.Edge(dest, alt);
                                queue.add(node);
                            }
                        }
                    }
                    ;
                }
            }
        }
        ;
        var count = gph.count;
        for (var i = 0; i < count; i++) {
            {
                if (dist[i] === 2147483647) {
                    console.info(" \n node id " + i + "  prev " + previous[i] + " distance : Unreachable");
                }
                else {
                    console.info(" node id " + i + "  prev " + previous[i] + " distance : " + dist[i]);
                }
            }
            ;
        }
    };
    GraphAM.prims = function (gph) {
        var previous = (function (s) { var a = []; while (s-- > 0)
            a.push(0); return a; })(gph.count);
        var dist = (function (s) { var a = []; while (s-- > 0)
            a.push(0); return a; })(gph.count);
        var source = 0;
        var visited = (function (s) { var a = []; while (s-- > 0)
            a.push(false); return a; })(gph.count);
        for (var i = 0; i < gph.count; i++) {
            {
                previous[i] = -1;
                dist[i] = 2147483647;
                visited[i] = false;
            }
            ;
        }
        dist[source] = 0;
        previous[source] = -1;
        var comp = new GraphAM.EdgeComparator();
        var queue = (new PriorityQueue(100, (comp)));
        var node = new GraphAM.Edge(source, 0);
        queue.add(node);
        while ((queue.isEmpty() !== true)) {
            {
                node = queue.peek();
                queue.remove();
                source = node.dest;
                visited[source] = true;
                for (var dest = 0; dest < gph.count; dest++) {
                    {
                        var cost = gph.adj[source][dest];
                        if (cost !== 0) {
                            var alt = cost;
                            if (dist[dest] > alt && visited[dest] === false) {
                                dist[dest] = alt;
                                previous[dest] = source;
                                node = new GraphAM.Edge(dest, alt);
                                queue.add(node);
                            }
                        }
                    }
                    ;
                }
            }
        }
        ;
        var count = gph.count;
        for (var i = 0; i < count; i++) {
            {
                if (dist[i] === 2147483647) {
                    console.info(" \n node id " + i + "  prev " + previous[i] + " distance : Unreachable");
                }
                else {
                    console.info(" node id " + i + "  prev " + previous[i] + " distance : " + dist[i]);
                }
            }
            ;
        }
    };
    GraphAM.main2 = function () {
        var gph = new GraphAM(9);
        gph.addUndirectedEdge(0, 1, 4);
        gph.addUndirectedEdge(0, 7, 8);
        gph.addUndirectedEdge(1, 2, 8);
        gph.addUndirectedEdge(1, 7, 11);
        gph.addUndirectedEdge(2, 3, 7);
        gph.addUndirectedEdge(2, 8, 2);
        gph.addUndirectedEdge(2, 5, 4);
        gph.addUndirectedEdge(3, 4, 9);
        gph.addUndirectedEdge(3, 5, 14);
        gph.addUndirectedEdge(4, 5, 10);
        gph.addUndirectedEdge(5, 6, 2);
        gph.addUndirectedEdge(6, 7, 1);
        gph.addUndirectedEdge(6, 8, 6);
        gph.addUndirectedEdge(7, 8, 7);
        gph.print();
        GraphAM.prims(gph);
        GraphAM.dijkstra(gph, 0);
    };
    GraphAM.main3 = function () {
        var gph = new GraphAM(9);
        gph.addUndirectedEdge(0, 2, 1);
        gph.addUndirectedEdge(1, 2, 5);
        gph.addUndirectedEdge(1, 3, 7);
        gph.addUndirectedEdge(1, 4, 9);
        gph.addUndirectedEdge(3, 2, 2);
        gph.addUndirectedEdge(3, 5, 4);
        gph.addUndirectedEdge(4, 5, 6);
        gph.addUndirectedEdge(4, 6, 3);
        gph.addUndirectedEdge(5, 7, 1);
        gph.addUndirectedEdge(6, 7, 7);
        gph.addUndirectedEdge(7, 8, 17);
        gph.print();
        GraphAM.prims(gph);
        GraphAM.dijkstra(gph, 1);
    };
    GraphAM.hamiltonianPathUtil = function (graph, path, pSize, added) {
        if (pSize === graph.count) {
            return true;
        }
        for (var vertex = 0; vertex < graph.count; vertex++) {
            {
                if (pSize === 0 || (graph.adj[path[pSize - 1]][vertex] === 1 && added[vertex] === 0)) {
                    path[pSize++] = vertex;
                    added[vertex] = 1;
                    if (GraphAM.hamiltonianPathUtil(graph, path, pSize, added))
                        return true;
                    pSize--;
                    added[vertex] = 0;
                }
            }
            ;
        }
        return false;
    };
    GraphAM.hamiltonianPath = function (graph) {
        var path = (function (s) { var a = []; while (s-- > 0)
            a.push(0); return a; })(graph.count);
        var added = (function (s) { var a = []; while (s-- > 0)
            a.push(0); return a; })(graph.count);
        if (GraphAM.hamiltonianPathUtil(graph, path, 0, added)) {
            console.info("Hamiltonian Path found :: ");
            for (var i = 0; i < graph.count; i++) {
                console.info(" " + path[i]);
            }
            return true;
        }
        console.info("Hamiltonian Path not found");
        return false;
    };
    GraphAM.hamiltonianCycleUtil = function (graph, path, pSize, added) {
        if (pSize === graph.count) {
            if (graph.adj[path[pSize - 1]][path[0]] === 1) {
                path[pSize] = path[0];
                return true;
            }
            else
                return false;
        }
        for (var vertex = 0; vertex < graph.count; vertex++) {
            {
                if (pSize === 0 || (graph.adj[path[pSize - 1]][vertex] === 1 && added[vertex] === 0)) {
                    path[pSize++] = vertex;
                    added[vertex] = 1;
                    if (GraphAM.hamiltonianCycleUtil(graph, path, pSize, added))
                        return true;
                    pSize--;
                    added[vertex] = 0;
                }
            }
            ;
        }
        return false;
    };
    GraphAM.hamiltonianCycle = function (graph) {
        var path = (function (s) { var a = []; while (s-- > 0)
            a.push(0); return a; })(graph.count + 1);
        var added = (function (s) { var a = []; while (s-- > 0)
            a.push(0); return a; })(graph.count);
        if (GraphAM.hamiltonianCycleUtil(graph, path, 0, added)) {
            console.info("Hamiltonian Cycle found :: ");
            for (var i = 0; i <= graph.count; i++) {
                console.info(" " + path[i]);
            }
            return true;
        }
        console.info("Hamiltonian Cycle not found");
        return false;
    };
    GraphAM.main4 = function () {
        var count = 5;
        var graph = new GraphAM(count);
        var adj = [[0, 1, 0, 1, 0], [1, 0, 1, 1, 0], [0, 1, 0, 0, 1], [1, 1, 0, 0, 1], [0, 1, 1, 1, 0]];
        for (var i = 0; i < count; i++) {
            for (var j = 0; j < count; j++) {
                if (adj[i][j] === 1)
                    graph.addDirectedEdge(i, j, 1);
                ;
            }
            ;
        }
        console.info("hamiltonianPath : " + GraphAM.hamiltonianPath(graph));
        console.info("hamiltonianCycle : " + GraphAM.hamiltonianCycle(graph));
        var graph2 = new GraphAM(count);
        var adj2 = [[0, 1, 0, 1, 0], [1, 0, 1, 1, 0], [0, 1, 0, 0, 1], [1, 1, 0, 0, 0], [0, 1, 1, 0, 0]];
        for (var i = 0; i < count; i++) {
            for (var j = 0; j < count; j++) {
                if (adj2[i][j] === 1)
                    graph2.addDirectedEdge(i, j, 1);
                ;
            }
            ;
        }
        console.info("hamiltonianPath :  " + GraphAM.hamiltonianPath(graph2));
        console.info("hamiltonianCycle :  " + GraphAM.hamiltonianCycle(graph2));
    };
    return GraphAM;
}());
GraphAM["__class"] = "GraphAM";
(function (GraphAM) {
    var Edge = /** @class */ (function () {
        function Edge(dst, cst) {
            if (this.dest === undefined)
                this.dest = 0;
            if (this.cost === undefined)
                this.cost = 0;
            this.dest = dst;
            this.cost = cst;
        }
        return Edge;
    }());
    GraphAM.Edge = Edge;
    Edge["__class"] = "GraphAM.Edge";
    var EdgeComparator = /** @class */ (function () {
        function EdgeComparator() {
        }
        EdgeComparator.prototype.compare = function (x, y) {
            if (x.cost < y.cost) {
                return -1;
            }
            if (x.cost > y.cost) {
                return 1;
            }
            return 0;
        };
        return EdgeComparator;
    }());
    GraphAM.EdgeComparator = EdgeComparator;
    EdgeComparator["__class"] = "GraphAM.EdgeComparator";
    EdgeComparator["__interfaces"] = ["Comparator"];
})(GraphAM || (GraphAM = {}));
GraphAM.main1();
GraphAM.main2();
GraphAM.main3();
GraphAM.main4();

File: Graph/GraphAM.java
import java.util.Arrays;
import java.util.Comparator;
import java.util.LinkedList;
import java.util.PriorityQueue;
import java.util.Stack;
import java.util.Queue;

public class GraphAM {
    int count;
    int[][] adj;

    GraphAM(int cnt) {
        count = cnt;
        adj = new int[count][count];
    }

    public void addDirectedEdge(int src, int dst, int cost) {
        adj[src][dst] = cost;
    }

    public void addUndirectedEdge(int src, int dst, int cost) {
        addDirectedEdge(src, dst, cost);
        addDirectedEdge(dst, src, cost);
    }

    public void print() {
        for (int i = 0; i < count; i++) {
            System.out.print("Vertex " + i + " is connected to : ");
            for (int j = 0; j < count; j++) {
                if (adj[i][j] != 0)
                    System.out.print("(" + j + ", " + adj[i][j] + ") ");
            }
            System.out.println("");
        }
    }

    public static void main1() {
        GraphAM graph = new GraphAM(4);
        graph.addUndirectedEdge(0, 1, 1);
        graph.addUndirectedEdge(0, 2, 1);
        graph.addUndirectedEdge(1, 2, 1);
        graph.addUndirectedEdge(2, 3, 1);
        graph.print();
    }

/*
Vertex 0 is connected to : (1, 1) (2, 1) 
Vertex 1 is connected to : (0, 1) (2, 1) 
Vertex 2 is connected to : (0, 1) (1, 1) (3, 1) 
Vertex 3 is connected to : (2, 1)
*/
    private static class Edge implements Comparable<Edge> {
        int src, dest, cost;

        public Edge(int s, int d, int c) {
            src = s;
            dest = d;
            cost = c;
        }

        public int compareTo(Edge compareEdge) {
            return this.cost - compareEdge.cost;
        }
    }

    public static void dijkstra(GraphAM gph, int source) {
        int[] previous = new int[gph.count];
        int[] dist = new int[gph.count];
        boolean[] visited = new boolean[gph.count];
        Arrays.fill(previous, -1);
        Arrays.fill(dist, Integer.MAX_VALUE); // infinite

        dist[source] = 0;
        previous[source] = -1;

        PriorityQueue<Edge> queue = new PriorityQueue<Edge>(100);
        Edge node = new Edge(source, source, 0);
        queue.add(node);

        while (queue.isEmpty() != true) {
            node = queue.peek();
            queue.remove();
            source = node.dest;
            visited[source] = true;
            for (int dest = 0; dest < gph.count; dest++) {
                int cost = gph.adj[source][dest];
                if (cost != 0) {
                    int alt = cost + dist[source];
                    if (dist[dest] > alt && visited[dest] == false) {

                        dist[dest] = alt;
                        previous[dest] = source;
                        node = new Edge(source, dest, alt);
                        queue.add(node);
                    }
                }
            }
        }

        int count = gph.count;
        for (int i = 0; i < count; i++) {
            if (dist[i] == Integer.MAX_VALUE) {
                System.out.println("node id " + i + "  prev " + previous[i] + " distance : Unreachable");
            } else {
                System.out.println("node id " + i + "  prev " + previous[i] + " distance : " + dist[i]);

            }
        }
    }

    public static void prims(GraphAM gph) {
        int[] previous = new int[gph.count];
        int[] dist = new int[gph.count];
        int source = 0;
        boolean[] visited = new boolean[gph.count];
        Arrays.fill(previous, -1);
        Arrays.fill(dist, Integer.MAX_VALUE); // infinite

        dist[source] = 0;
        previous[source] = -1;
        PriorityQueue<Edge> queue = new PriorityQueue<Edge>(100);
        Edge node = new Edge(source, source, 0);
        queue.add(node);

        while (queue.isEmpty() != true) {
            node = queue.peek();
            queue.remove();
            source = node.dest;
            visited[source] = true;
            for (int dest = 0; dest < gph.count; dest++) {
                int cost = gph.adj[source][dest];
                if (cost != 0) {
                    if (dist[dest] > cost && visited[dest] == false) {
                        dist[dest] = cost;
                        previous[dest] = source;
                        node = new Edge(source, dest, cost);
                        queue.add(node);
                    }
                }
            }
        }

        int count = gph.count;
        for (int i = 0; i < count; i++) {
            if (dist[i] == Integer.MAX_VALUE) {
                System.out.println("node id " + i + "  prev " + previous[i] + " distance : Unreachable");
            } else {
                System.out.println("node id " + i + "  prev " + previous[i] + " distance : " + dist[i]);
            }
        }
    }

    public static void main2() {
        GraphAM gph = new GraphAM(9);
        gph.addUndirectedEdge(0, 1, 4);
        gph.addUndirectedEdge(0, 7, 8);
        gph.addUndirectedEdge(1, 2, 8);
        gph.addUndirectedEdge(1, 7, 11);
        gph.addUndirectedEdge(2, 3, 7);
        gph.addUndirectedEdge(2, 8, 2);
        gph.addUndirectedEdge(2, 5, 4);
        gph.addUndirectedEdge(3, 4, 9);
        gph.addUndirectedEdge(3, 5, 14);
        gph.addUndirectedEdge(4, 5, 10);
        gph.addUndirectedEdge(5, 6, 2);
        gph.addUndirectedEdge(6, 7, 1);
        gph.addUndirectedEdge(6, 8, 6);
        gph.addUndirectedEdge(7, 8, 7);
        //gph.print();
        prims(gph);
        //dijkstra(gph, 0);
    }
/*
Vertex 0 is connected to : (1, 4) (7, 8) 
Vertex 1 is connected to : (0, 4) (2, 8) (7, 11) 
Vertex 2 is connected to : (1, 8) (3, 7) (5, 4) (8, 2) 
Vertex 3 is connected to : (2, 7) (4, 9) (5, 14) 
Vertex 4 is connected to : (3, 9) (5, 10) 
Vertex 5 is connected to : (2, 4) (3, 14) (4, 10) (6, 2) 
Vertex 6 is connected to : (5, 2) (7, 1) (8, 6) 
Vertex 7 is connected to : (0, 8) (1, 11) (6, 1) (8, 7) 
Vertex 8 is connected to : (2, 2) (6, 6) (7, 7)  


node id 0  prev -1 distance : 0
node id 1  prev 0 distance : 4
node id 2  prev 5 distance : 4
node id 3  prev 2 distance : 7
node id 4  prev 3 distance : 9
node id 5  prev 6 distance : 2
node id 6  prev 7 distance : 1
node id 7  prev 0 distance : 8
node id 8  prev 2 distance : 2

node id 0  prev -1 distance : 0
node id 1  prev 0 distance : 4
node id 2  prev 1 distance : 12
node id 3  prev 2 distance : 19
node id 4  prev 5 distance : 21
node id 5  prev 6 distance : 11
node id 6  prev 7 distance : 9
node id 7  prev 0 distance : 8
node id 8  prev 2 distance : 14

*/

    public static void main3() {
        GraphAM gph = new GraphAM(9);
        gph.addUndirectedEdge(0, 2, 1);
        gph.addUndirectedEdge(1, 2, 5);
        gph.addUndirectedEdge(1, 3, 7);
        gph.addUndirectedEdge(1, 4, 9);
        gph.addUndirectedEdge(3, 2, 2);
        gph.addUndirectedEdge(3, 5, 4);
        gph.addUndirectedEdge(4, 5, 6);
        gph.addUndirectedEdge(4, 6, 3);
        gph.addUndirectedEdge(5, 7, 1);
        gph.addUndirectedEdge(6, 7, 7);
        gph.addUndirectedEdge(7, 8, 17);
        gph.print();
        prims(gph);
        dijkstra(gph, 1);
    }
/*
Vertex 0 is connected to : (2, 1) 
Vertex 1 is connected to : (2, 5) (3, 7) (4, 9) 
Vertex 2 is connected to : (0, 1) (1, 5) (3, 2) 
Vertex 3 is connected to : (1, 7) (2, 2) (5, 4) 
Vertex 4 is connected to : (1, 9) (5, 6) (6, 3) 
Vertex 5 is connected to : (3, 4) (4, 6) (7, 1) 
Vertex 6 is connected to : (4, 3) (7, 7) 
Vertex 7 is connected to : (5, 1) (6, 7) (8, 17) 
Vertex 8 is connected to : (7, 17)

node id 0  prev -1 distance : 0
node id 1  prev 2 distance : 5
node id 2  prev 0 distance : 1
node id 3  prev 2 distance : 2
node id 4  prev 5 distance : 6
node id 5  prev 3 distance : 4
node id 6  prev 4 distance : 3
node id 7  prev 5 distance : 1
node id 8  prev 7 distance : 17

node id 0  prev 2 distance : 6
node id 1  prev -1 distance : 0
node id 2  prev 1 distance : 5
node id 3  prev 1 distance : 7
node id 4  prev 1 distance : 9
node id 5  prev 3 distance : 11
node id 6  prev 4 distance : 12
node id 7  prev 5 distance : 12
node id 8  prev 7 distance : 29
*/
    public static boolean hamiltonianPathUtil(GraphAM graph, int path[], int pSize, int added[]) {
        // Base case full length path is found
        if (pSize == graph.count) {
            return true;
        }
        for (int vertex = 0; vertex < graph.count; vertex++) {
            // There is an edge from last element of path and next vertex
            // and the next vertex is not already included in the path.
            if (pSize == 0 || (graph.adj[path[pSize - 1]][vertex] == 1 && added[vertex] == 0)) {
                path[pSize++] = vertex;
                added[vertex] = 1;
                if (hamiltonianPathUtil(graph, path, pSize, added))
                    return true;
                // backtracking
                pSize--;
                added[vertex] = 0;
            }
        }
        return false;
    }

    public static boolean hamiltonianPath(GraphAM graph) {
        int[] path = new int[graph.count];
        int[] added = new int[graph.count];

        if (hamiltonianPathUtil(graph, path, 0, added)) {
            System.out.print("Hamiltonian Path found :: ");
            for (int i = 0; i < graph.count; i++)
                System.out.print(" " + path[i]);
            System.out.println("");
            return true;
        }

        System.out.println("Hamiltonian Path not found");
        return false;
    }

    public static boolean hamiltonianCycleUtil(GraphAM graph, int path[], int pSize, int added[]) {
        // Base case full length path is found
        // this last check can be modified to make it a path.
        if (pSize == graph.count) {
            if (graph.adj[path[pSize - 1]][path[0]] == 1) {
                path[pSize] = path[0];
                return true;
            } else
                return false;
        }
        for (int vertex = 0; vertex < graph.count; vertex++) {
            // there is a path from last element and next vertex
            if (pSize == 0 || (graph.adj[path[pSize - 1]][vertex] == 1 && added[vertex] == 0)) {
                path[pSize++] = vertex;
                added[vertex] = 1;
                if (hamiltonianCycleUtil(graph, path, pSize, added))
                    return true;
                // backtracking
                pSize--;
                added[vertex] = 0;
            }
        }
        return false;
    }

    public static boolean hamiltonianCycle(GraphAM graph) {
        int[] path = new int[graph.count + 1];
        int[] added = new int[graph.count];
        if (hamiltonianCycleUtil(graph, path, 0, added)) {
            System.out.print("Hamiltonian Cycle found :: ");
            for (int i = 0; i <= graph.count; i++)
                System.out.print(" " + path[i]);
            System.out.println("");
            return true;
        }
        System.out.println("Hamiltonian Cycle not found");
        return false;
    }

    public static void main4() {
        int count = 5;
        GraphAM graph = new GraphAM(count);
        int[][] adj = { 
            { 0, 1, 0, 1, 0 }, 
            { 1, 0, 1, 1, 0 }, 
            { 0, 1, 0, 0, 1 }, 
            { 1, 1, 0, 0, 1 }, 
            { 0, 1, 1, 1, 0 } };

        for (int i = 0; i < count; i++)
            for (int j = 0; j < count; j++)
                if (adj[i][j] == 1)
                    graph.addDirectedEdge(i, j, 1);
        System.out.println("hamiltonianPath : " + hamiltonianPath(graph));

        GraphAM graph2 = new GraphAM(count);
        int[][] adj2 = { 
            { 0, 1, 0, 1, 0 }, 
            { 1, 0, 1, 1, 0 }, 
            { 0, 1, 0, 0, 1 }, 
            { 1, 1, 0, 0, 0 },
            { 0, 1, 1, 0, 0 } };
        for (int i = 0; i < count; i++)
            for (int j = 0; j < count; j++)
                if (adj2[i][j] == 1)
                    graph2.addDirectedEdge(i, j, 1);

        System.out.println("hamiltonianPath :  " + hamiltonianPath(graph2));
    }
/*
Hamiltonian Path found ::  0 1 2 4 3
hamiltonianPath : true

Hamiltonian Path found ::  0 3 1 2 4
hamiltonianPath :  true
 */
    public static void main5() {
        int count = 5;
        GraphAM graph = new GraphAM(count);
        int[][] adj = { 
            { 0, 1, 0, 1, 0 }, 
            { 1, 0, 1, 1, 0 }, 
            { 0, 1, 0, 0, 1 }, 
            { 1, 1, 0, 0, 1 }, 
            { 0, 1, 1, 1, 0 } };

        for (int i = 0; i < count; i++)
            for (int j = 0; j < count; j++)
                if (adj[i][j] == 1)
                    graph.addDirectedEdge(i, j, 1);
        System.out.println("hamiltonianCycle : " + hamiltonianCycle(graph));

        GraphAM graph2 = new GraphAM(count);
        int[][] adj2 = { 
            { 0, 1, 0, 1, 0 }, 
            { 1, 0, 1, 1, 0 }, 
            { 0, 1, 0, 0, 1 }, 
            { 1, 1, 0, 0, 0 },
            { 0, 1, 1, 0, 0 } };
        for (int i = 0; i < count; i++)
            for (int j = 0; j < count; j++)
                if (adj2[i][j] == 1)
                    graph2.addDirectedEdge(i, j, 1);

        System.out.println("hamiltonianCycle :  " + hamiltonianCycle(graph2));
    }

/*
Hamiltonian Cycle found ::  0 1 2 4 3 0
hamiltonianCycle : true

Hamiltonian Cycle not found
hamiltonianCycle :  false
*/
    public static void main(String[] args) {
            //main1();
            main2(); 
            //main3(); 
            //main4(); 
    }
}